### PR TITLE
LIVE-1760 Celo staking in LLM

### PR DESCRIPTION
**Pull request by haammar-ledger**
### 📝 Description

Support of Celo staking was added into LLD with https://github.com/LedgerHQ/ledger-live/pull/818, this adds support for it in LLM.

### ❓ Context

- **Impacted projects**: `ledger-live-mobile` 
- **Linked resource(s)**: https://ledgerhq.atlassian.net/browse/LIVE-1760 

### ✅ Checklist

- [x] **Test coverage** 
- [x] **Atomic delivery** 
- [x] **No breaking changes**

### CODE REVIEW

File: libs/ledger-live-common/src/families/celo/specs.ts [hunk 1]
```diff
@@ -1,14 +1,30 @@
...
-const maxAccount = 3;
+const send50PercentMutation = createSend50PercentMutation();
+const sendMaxMutation = createSendMaxMutation();
+const registerAccountMutation = createRegisterAccountMutation();
+const unlockMutation = createUnlockMutation();
+const lockMutation = createLockMutation();
+const voteMutation = createVoteMutation();
+const activateVoteMutation = createActivateVoteMutation();
+const revokeVoteMutation = createRevokeVoteMutation();
+const withdrawMutation = createWithdrawMutation();
```
Comment (gre) [inline libs/ledger-live-common/src/families/celo/specs.ts:27]: what is the idea behind these functions? as celo would be the first to adopt this pattern, i'm not sure this helps readability and future maintenance. also what happens to the "maxAccount" shared value then?

alternatively you could define all the mutation as local vailable in that same file, and then compose your mutations:[...] as you did 🤔

**Review by gre — DISMISSED** [review]
small comment on the bot changes.

also please increase the "maxAccount" to a higher number because i don't feel we currently test enough cases on the bot. how long is the "lock/unlock" system in Celo? there needs to be enough account so we have a high chance to test ALL possible mutations, and here I actually see more mutation that there are accounts !

File: libs/ledgerjs/packages/cryptoassets/package.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@ledgerhq/cryptoassets",
-  "version": "6.33.0",
+  "version": "6.34.0-next.0",
```
Comment (gre) [inline libs/ledgerjs/packages/cryptoassets/package.json:3]: this shouldn't happen in a PR.

File: libs/ledgerjs/packages/hw-app-eth/package.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@ledgerhq/hw-app-eth",
-  "version": "6.29.5",
+  "version": "6.29.6-next.0",
```
Comment (gre) [inline libs/ledgerjs/packages/hw-app-eth/package.json:3]: same

File: libs/ledger-live-common/src/families/celo/specs/createActivateVoteMutation.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
+import invariant from "invariant";
+import { getCryptoCurrencyById, parseCurrencyUnit } from "../../../currencies";
+import { MutationSpec } from "../../../bot/types";
+import type { CeloAccount, Transaction } from "../types";
+
+const currency = getCryptoCurrencyById("celo");
+const minimalAmount = parseCurrencyUnit(currency.units[0], "0.001");
+
+export const createActivateVoteMutation = (): MutationSpec<Transaction> => ({
```
Comment (gre) [inline libs/ledger-live-common/src/families/celo/specs/createActivateVoteMutation.ts:9]: ⚠️  there are not even a single test on this mutation

File: libs/ledger-live-common/src/families/celo/specs/createLockMutation.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
+import invariant from "invariant";
+import { getCryptoCurrencyById, parseCurrencyUnit } from "../../../currencies";
+import { MutationSpec } from "../../../bot/types";
+import type { CeloAccount, Transaction } from "../types";
+
+const currency = getCryptoCurrencyById("celo");
+const minimalAmount = parseCurrencyUnit(currency.units[0], "0.005");
+
+export const createLockMutation = (): MutationSpec<Transaction> => ({
```
Comment (gre) [inline libs/ledger-live-common/src/families/celo/specs/createLockMutation.ts:9]: ⚠️  there are not even a single test on this mutation

File: libs/ledger-live-common/src/families/celo/specs/createRegisterAccountMutation.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
+import invariant from "invariant";
+import { getCryptoCurrencyById, parseCurrencyUnit } from "../../../currencies";
+import { MutationSpec } from "../../../bot/types";
+import type { CeloAccount, Transaction } from "../types";
+
+const currency = getCryptoCurrencyById("celo");
+const minimalAmount = parseCurrencyUnit(currency.units[0], "0.001");
+
+export const createRegisterAccountMutation = (): MutationSpec<Transaction> => ({
```
Comment (gre) [inline libs/ledger-live-common/src/families/celo/specs/createRegisterAccountMutation.ts:9]: ⚠️  there are not even a single test on this mutation

File: libs/ledger-live-common/src/families/celo/specs/createRevokeVoteMutation.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
+import invariant from "invariant";
+import { getCryptoCurrencyById, parseCurrencyUnit } from "../../../currencies";
+import { MutationSpec } from "../../../bot/types";
+import type { CeloAccount, Transaction } from "../types";
+
+const currency = getCryptoCurrencyById("celo");
+const minimalAmount = parseCurrencyUnit(currency.units[0], "0.001");
+
+export const createRevokeVoteMutation = (): MutationSpec<Transaction> => ({
```
Comment (gre) [inline libs/ledger-live-common/src/families/celo/specs/createRevokeVoteMutation.ts:9]: ⚠️  there are not even a single test on this mutation

File: libs/ledger-live-common/src/families/celo/specs/createSendMutation.ts [hunk 8]
```diff
@@ -0,0 +1,44 @@
+import invariant from "invariant";
+import { getCryptoCurrencyById, parseCurrencyUnit } from "../../../currencies";
+import { pickSiblings } from "../../../bot/specs";
+import { MutationSpec } from "../../../bot/types";
+import type { Transaction } from "../types";
+
+const maxAccount = 3;
+const currency = getCryptoCurrencyById("celo");
+const minimalAmount = parseCurrencyUnit(currency.units[0], "0.001");
+
```
Comment (gre) [inline libs/ledger-live-common/src/families/celo/specs/createSendMutation.ts:10]: ⚠️  there are not even a single test on these mutations

**Review by gre — COMMENTED** [review]
also, there are missing tests on all mutation. i'm not sure what we really test on these mutation (apart from the fact that it "broadcasts"), we don't verify the impact of the mutation on the account, so the bot spec is currently quite weak and need improvmeents

File: apps/ledger-live-mobile/src/icons/Loader.js [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+      />
+    </Svg>
+  );
+}
```
Comment (gre) [inline apps/ledger-live-mobile/src/icons/Loader.js:23]: need to `memo` that component

Comment (gre) [reply apps/ledger-live-mobile/src/icons/Loader.js:23]: cc @LFBarreto

File: libs/ledgerjs/packages/cryptoassets/package.json (hunk 2 above)
Comment (haammar-ledger) [reply libs/ledgerjs/packages/cryptoassets/package.json]: Probably caused by a bad merge. Fixed now.

**Review by henrily-ledger — APPROVED** [review]

File: libs/ledger-live-common/src/families/celo/specs/createActivateVoteMutation.ts (hunk 4 above)
Comment (andyhass) [reply libs/ledger-live-common/src/families/celo/specs/createActivateVoteMutation.ts]: @gre  correct me if I'm wrong, but there is an inherit test that's run on each mutation that verifies that the transaction was broadcasted successfully - that said, the most critical functionality is, in fact, tested.  This pattern exists in Cosmos bot tests, among others.

File: libs/ledger-live-common/src/families/celo/specs.ts (hunk 1 above)
Comment (andyhass) [reply libs/ledger-live-common/src/families/celo/specs.ts]: Hey @gre - I can explain some of the decisions here.  

Regarding the factory functions that return mutations, I put these in separate files for readability and maintainability.  Breaking apart the mutations into separate files was, to me, a superior to having a 1000 line spec file.  

Regarding sharing variables, that contradicts the direction I was given - @haammar-ledger indicated that each mutation _should_ be able to live on it's own and not be dependent on other mutations (see discussion in Discord [here](https://discord.com/channels/885256081289379850/950694044802097172/1002249261800493107)).

Comment (haammar-ledger) [reply libs/ledger-live-common/src/families/celo/specs.ts]: Readability is a matter of taste :slightly_smiling_face: I also don't mind to have this code split like that.
LGTM, closing this thread, feel free to reopen.

File: libs/ledger-live-common/src/families/celo/specs/createActivateVoteMutation.ts (hunk 4 above)
Comment (haammar-ledger) [reply libs/ledger-live-common/src/families/celo/specs/createActivateVoteMutation.ts]: @andyhass Thanks for your feedback.
What @gre mentions here is the point of adding a `test:` field within every mutation, or even at the global spec level.
When there is no test defined, what the bot tests is indeed that the broadcast succeeds, which is the minimum, but there are often many additional checks that can be done around the transaction to validate more things.
Probably not all of our specs do it (yet), but you can check many examples for instance the one for cosmos does have additional tests defined:
- https://github.com/LedgerHQ/ledger-live/blob/2da52a7b570bb4e20ba2d40735e1d5ef082c3f31/libs/ledger-live-common/src/families/cosmos/specs.ts#L68
- https://github.com/LedgerHQ/ledger-live/blob/2da52a7b570bb4e20ba2d40735e1d5ef082c3f31/libs/ledger-live-common/src/families/cosmos/specs.ts#L102
etc.

**Pull request merged**